Source/Mosa.DeviceDriver/ISA/ACPI.cs

```diff
@@ -0,0 +1,251 @@
+		public uint PM1aControlBlock;
+		public uint PM1bControlBlock;
+		public uint PM2ControlBlock;
+		public uint PMTimerBlock;
+		public uint GPE0Block;
+		public uint GPE1Block;
+		public byte PM1EventLength;
+		public byte PM1ControlLength;
+		public byte PM2ControlLength;
+		public byte PMTimerLength;
+		public byte GPE0Length;
+		public byte GPE1Length;
+		public byte GPE1Base;
+		public byte CStateControl;
+		public ushort WorstC2Latency;
+		public ushort WorstC3Latency;
+		public ushort FlushSize;
+		public ushort FlushStride;
+		public byte DutyOffset;
+		public byte DutyWidth;
+		public byte DayAlarm;
+		public byte MonthAlarm;
+		public byte Century;
+
+		// Reserved in ACPI 1.0; used since ACPI 2.0+
+		public ushort BootArchitectureFlags;
+
+		public byte Reserved2;
+		public uint Flags;
+
+		// 12 byte structure; see below for details
+		public GenericAddressStructure ResetReg;
+
+		public byte ResetValue;
+		public fixed byte Reserved3[3];
+
+		// 64bit pointers - Available on ACPI 2.0+
+		public ulong X_FirmwareControl;
+		public ulong X_Dsdt;
+
+		public GenericAddressStructure X_PM1aEventBlock;
+		public GenericAddressStructure X_PM1bEventBlock;
+		public GenericAddressStructure X_PM1aControlBlock;
+		public GenericAddressStructure X_PM1bControlBlock;
+		public GenericAddressStructure X_PM2ControlBlock;
+		public GenericAddressStructure X_PMTimerBlock;
+		public GenericAddressStructure X_GPE0Block;
+		public GenericAddressStructure X_GPE1Block;
+	}
+
+	[StructLayout(LayoutKind.Sequential, Pack = 1)]
+	public unsafe struct GenericAddressStructure
+	{
+		public byte AddressSpace;
+		public byte BitWidth;
+		public byte BitOffset;
+		public byte AccessSize;
+		public ulong Address;
+	}
+
+	[StructLayout(LayoutKind.Sequential, Pack = 1)]
+	public unsafe struct RSDT
+	{
+		public ACPISDTHeader h;
+		public fixed uint PointerToOtherSDT[8]; // This is problematic, we need a way to statically initialize this array's size with h.Length and stuff
+
+		/*public void Init()
+        {
+            fixed (uint* ptr = new uint[(int)((h.Length - sizeof(ACPISDTHeader)) / 4)])
+                PointerToOtherSDT = ptr;
+        }*/
+	}
+
+	public unsafe class ACPI : BaseDeviceDriver
+	{
+		private RSDPDescriptor* Descriptor;
+		private FADT* FADT;
+		private RSDT* RSDT;
+
+		private short SLP_TYPa;
+		private short SLP_TYPb;
+		private short SLP_EN;
+
+		private BaseIOPortWrite SMI_CommandPort;
+		private BaseIOPortWrite PM1aControlBlock;
+		private BaseIOPortWrite PM1bControlBlock;
+
+		public override void Initialize()
+		{
+			Device.Name = "ACPI";
+
+			Pointer rsdpPtr;
+
+			for (uint addr = 0x000E0000; addr <= 0x000FFFFF; addr += 8)
+			{
+				rsdpPtr = (Pointer)addr;
+
+				if (rsdpPtr.Load64() == 0x2052545020445352) // 'RSD PTR '
+					Descriptor = (RSDPDescriptor*)rsdpPtr;
+			}
+
+			if (Descriptor == null)
+				return;
+
+			RSDT = (RSDT*)HAL.GetPhysicalMemory((Pointer)Descriptor->RsdtAddress, 0xFFFF).Address;
```

Comment:
I prefer not using unsafe code and using this pattern instead: 

https://github.com/mosa/MOSA-Project/blob/master/Source/Mosa.Kernel.BareMetal/MultibootSpecification/MultibootV1.cs

Reply:
The problem is that then the code will be huge, and the offsets of ACPI tables aren't specified anywhere on the osdev wiki, so it's close to impossible.